Cache API tests & unifying behavior

waynew opened the pull request:
### What does this PR do?

Adds API tests for caching, making it relatively trivial to add new cache backends that conform to the default/localfs cache behavior. It also fixes a few inconsistencies (in particular the missing `updated` functionality for a few of the backends.)

This also functions as regression tests since it will ensure that any changes to cache backends that alter existing cache behavior will be exposed.

This will also be super useful when we upgrade mysql/redis/consul/etcd libraries and databases.

### What issues does this PR fix or reference?

Closes #61081
Closes #60273
Closes #60272 

### Previous Behavior

Cache backends were somewhat inconsistent. Also almost completely untested.

### New Behavior

All current cache backends behave identically and completely like the localfs cache backend

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs (Not sure if we want to document *more* here)
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

waynew committed:
Add tests for cache backends

waynew committed:
Tests run against everything!

waynew committed:
Get consul passing cache tests

waynew committed:
now mysql tests also pass

waynew committed:
comments

waynew committed:
add mysql cache tests

waynew committed:
Ensure consul cache is flushing timestamp keys

waynew committed:
Ensure etcd cache is flushing timestamp keys

waynew committed:
Update redis cache api to conform to localfs

There was a bug in some of the lookups where it was looking at sub banks
instead of the banks containing the leaf/key nodes. That was fixed. Also
restored the original behaviors, only adding the updated functionality,
namely adding timestamp key/values to redis.

waynew committed:
Fix etcd path inconsistency

Previously on X-Files. Wait nope. Previously on etcd cache files...
:joy: if a path were 1/2/3 then etcd would break with IndexError, but
when it was 1/2/3/4/5 then instead of returning the "file" (key?) 5 it
would return `4/5` as the name. Which is probably not correct, and
definitely inconsistent with localfs (and every other cache module).

Now it will always only return the final segment of the cache (e.g. 7
for 1/2/3/4/5/6/7 and 5 for 1/2/5)

waynew committed:
Remove comments and cleanup test setup

Comments were outdated, now they're gone or fixed. Also shifted around
some of the test setup to make it a bit cleaner.

waynew committed:
Add localfs backed memcache to the tests

Fixed an inconsistency when flushing banks with None key. Also added
to the test run. We may want to add other backing caches via alternate
fixtures?

waynew committed:
Bad serial

waynew committed:
No this is wrong

waynew committed:
Probably the right lazy loader fix

waynew committed:
Restore tested mysql cache functionality

Added a force_reconnect function because I couldn't figure out how to
remove the client from the context any other way. There may be a better
way to do that, but there will probably be some loader changes anyway.

waynew commented on salt/cache/etcd_cache.py:
```diff
@@ -182,7 +195,10 @@ def _walk(r):
+        if r.key.endswith(_tstamp_suffix):
+            return []
+        else:
+            return [r.key.rsplit("/", 1)[-1]]
```
Previously ~on X-Files~ in this module we required a specific nesting - in particular IIRC it wanted `1/2/3` for example. This is now fixed to where arbitrary nesting will work correctly in etcd. Including something absolutely ridiculous (see the tests for an example).

Ch3LL replied:
This looks like we would be breaking backwards capabilities. Would this break users existing setups if using the etcd cache?

waynew replied:
I'm not sure I follow - before it did /not/ allow arbitrary nesting (and localfs used os.walk, IIRC)

It only functioned correctly with precisely 3 levels of nesting - anything else would produce incorrect results at other levels of nesting.

I can double check, but... It was definitely inconsistent behavior, and only worked by accident, basically.

waynew replied:
First off: Thank you for the challenge! It's actually be super helpful for me to go back and double check this :+1: 

I was pretty confident that this is doing the right thing but now I'm absolutely positive.

Consider the default salt setup and run the following:

```
salt-run cache.store yourminion thing "some data goes here"
salt-run cache.list yourminion
# output:
# - thing
salt-run cache.fetch yourminion thing
# output:
# some data goes here
```

Now change the master config:

```
cache: etcd
etcd.host: 172.17.0.2   # my docker container
etcd.port: 2379
etcd.protocol: http
etcd.path_prefix: /foo/bar/bang
```

re-run the same commands and the list one will produce the output:

```
bang/yourminion/thing
```

Which is uh... not the same as localfs. From that point alone, I'd say that the existing behavior is buggy.

As far as tripping up anyone who's currently using etcd caching, I don't *think* that there will be any issues, because will store and retrieve the values the same, but that's basically because from etcd's perspective `'/this' + '/that/path'` is no different from `'/this/that' + '/path'`. In other words I'm pretty sure it's just by accident that it works at all. 

So we're definitely correcting a bug with the way that list works with the runner. I'm guessing the only reason we haven't seen any problems is because nobody ever hit this. Though [this issue might have something to do with it](https://github.com/saltstack/salt/issues/56175).

After all this poking around, I'm also fairly confident that we're not going to blow anything up - I spun up this container running the master branch version of salt and got some data cached, then I changed to this version and from what I could see by running curl commands and trying to fetch cache data there weren't any changes between what was stored. It's *possible* that I missed something, and I wouldn't be against putting another note of caution in the changelog if we think it'd be helpful.

waynew commented on tests/pytests/functional/cache/test_cache.py:
```diff
@@ -0,0 +1,526 @@
+        # and is at least the current behavior. If we want to change that
+        # this test should change. Or be removed altogether.
+        # TODO: this should actually not raise. Not sure if there's a test that we can do here... or just call the code which will fail if there's actually an exception. -W. Werner, 2021-09-28
+        pytest.skip(
+            "Skipping for now - etcd, redis, and mysql do not raise. Should ensure all backends behave consistently"
+        )
```
This is another accidental API. It's only the localfs cache that raises but I'm pretty sure it should not.

waynew committed:
remove silly files

Saltfile never should have been added in the first place and the
integration test wasn't really necessary, functional was better/lighter.

s0undt3ch commented on tests/pytests/functional/cache/test_cache.py:
```diff
@@ -0,0 +1,526 @@
+        check_ports=[consul_port],
+        container_run_kwargs={"ports": {"8500/tcp": consul_port}},
+    )
+    with container.started() as factory:
+        # TODO: May want to do the same thing for redis to ensure that service is up & running
+        # TODO: THIS IS HORRIBLE. THERE ARE BETTER WAYS TO DETECT SERVICE IS UP -W. Werner, 2021-10-12
+
+        timer = Timer(timeout=10)
+        sleeptime = 0.1
+        while not timer.expired:
+            try:
+                with socket.create_connection(
+                    ("localhost", consul_port), timeout=1
+                ) as cli:
+                    cli.send(b"GET /v1/kv/fnord HTTP/1.1\n\n")
+                    cli.recv(2048)
+                    break
+            except ConnectionResetError as e:
+                if e.errno == 104:
+                    time.sleep(sleeptime)
+                    sleeptime += sleeptime
+        else:
+            assert False, "Timer expired before connecting to consul"
+        yield factory
```
Check connection to ports not enough?
https://github.com/saltstack/pytest-salt-factories/blob/master/src/saltfactories/daemons/container.py#L64

waynew replied:
I don't think so - the ports can be open but the service isn't actually ready. This seems like the best option then, currently.

s0undt3ch replied:
Maybe we should investigating supporting making requests against services and assert on the responses if/when the service is up....

waynew committed:
Fix redis_cache cluster mode import

Fixes #60272 - the import statement was totally incorrect. Now it will
actually attempt to connect to a redis cluster.

waynew committed:
payload should string-ify all the data

waynew committed:
Fix unit tests for parametrization

From what I can tell, previously the query was being built in Python
instead of using mysql to add the parameters, and this probably horked
up the handling of binary strings.

Unless some other version of the DB will be problematic...

waynew committed:
Fix missing docstring complaints

waynew committed:
Merge remote-tracking branch 'salt/master' into tc/cache_tests

s0undt3ch commented on salt/cache/__init__.py:
```diff
@@ -345,5 +345,10 @@ def store(self, bank, key, data):
     def flush(self, bank, key=None):
-        self.storage.pop((bank, key), None)
+        if key is None:
+            for bank, key in tuple(self.storage):
+                if bank == bank:
+                    self.storage.pop((bank, key))
```
```suggestion
            for _bank, _key in tuple(self.storage):
                if _bank == bank:
                    self.storage.pop((bank, key))
```
?
You masked the function `bank` variable

waynew replied:
:facepalm: nice catch

waynew replied:
Got this one now

s0undt3ch commented on salt/cache/consul.py:
```diff
@@ -166,22 +186,37 @@ def list_(bank):
-                "There was an error getting the key, {}: {}".format(c_key, exc)
-            )
-        return value is not None
+    key = key or ""
+    try:
+        c_key = "{}/{}".format(bank, key)
+        _, value = api.kv.get(c_key, keys=True)
+    except Exception as exc:  # pylint: disable=broad-except
+        raise SaltCacheError(
+            "There was an error getting the key, {}: {}".format(c_key, exc)
+        )
+    return value is not None
```
```suggestion
    try:
        c_key = "{}/{}".format(bank, key or "")
        _, value = api.kv.get(c_key, keys=True)
    except Exception as exc:  # pylint: disable=broad-except
        raise SaltCacheError(
            "There was an error getting the key, {}: {}".format(c_key, exc)
        )
    return value is not None
```
This suggestion is just personal taste, since we don't use the variable anywhere else in the function body

waynew replied:
done

s0undt3ch commented on salt/cache/mysql_cache.py:
```diff
@@ -90,17 +96,30 @@ def __virtual__():
+    try:
+        __context__.pop("mysql_client")
+    except KeyError:
+        pass
```
Just `__context__.pop("mysql_client", None)` to avoid the `try/except`...

s0undt3ch commented on salt/cache/redis_cache.py:
```diff
@@ -203,6 +213,9 @@ def __virtual__():
 
 
 def init_kwargs(kwargs):
+    """
+    Effectively a noop. Return an empty dictionary.
+    """
     return {}
```
Do we really need this function? Since it just returns an empty dictionary....

Just asking, since I didn't dig deep enough as to why.

waynew replied:
Unfortunately... yeah. It's part of the cache signature :facepalm:

s0undt3ch commented on salt/cache/redis_cache.py:
```diff
@@ -275,13 +291,21 @@ def _get_bank_redis_key(bank):
 
+def _get_timestamp_key(bank, key):
+    opts = _get_redis_keys_opts()
+    return f"{opts['timestamp_prefix']}{opts['separator']}{bank}/{key}"
```
No f strings yet, remember Debian 9
:cry:

waynew replied:
Bah! :cry:

waynew replied:
Done... but not happy about it :trollface:

s0undt3ch commented on tasks/docstrings.py:
```diff
@@ -28,7 +28,6 @@
     "salt/beacons/salt_monitor.py": ["validate", "beacon"],
     "salt/beacons/watchdog.py": ["close", "to_salt_event"],
     "salt/cache/localfs.py": ["get_storage_id", "init_kwargs"],
-    "salt/cache/redis_cache.py": ["init_kwargs"],
```
:bow:

s0undt3ch commented on tests/pytests/functional/cache/test_cache.py:
```diff
@@ -0,0 +1,526 @@
+    return get_unused_localhost_port()
+
+
+# TODO: We should probably be building our own etcd docker image - fine to base it off of this one (or... others) -W. Werner, 2021-07-27
```
Probably we should consider maintaining our own? If there are not official images....

waynew replied:
I'm :+1: for that. Is that something you've been known to do? Is that something that I can do? It probably makes sense to copy from whatever official images there are anyway, similar to what we do with our golden images.

s0undt3ch replied:
Yeah, probably a copy of upstream would suffice

s0undt3ch commented on tests/pytests/functional/cache/test_cache.py:
```diff
@@ -0,0 +1,524 @@
+def ensure_deps(states):
+    installation_result = states.pip.installed(
+        name="fnord",
+        pkgs=["python-etcd", "redis", "redis-py-cluster", "python-consul", "pymysql"],
```
While I see what you're trying to do here, keep requirements in a single place, this won't pin to specific CI versions, which, from time to time, might break.
Why aren't we adding these requirements to the CI requirements files?

waynew replied:
They could (should?) probably go there as well.

Actually what I was going for is to mandate that this test runs, instead of getting skipped because requirements aren't installed.

Though, now that I'm thinking about it... for these modules I think it's actually a better approach to ensure that whatever is the latest dependency is tested, so that if (when) something breaks we get advanced notice of it. The alternative is for the documentation for these modules to specify that they've only been tested with a particular version of the dependencies.

s0undt3ch commented on tests/pytests/functional/cache/test_cache.py:
```diff
@@ -0,0 +1,524 @@
+    opts["cache"] = "localfs"
+    cache = salt.cache.factory(opts)
+    yield cache
+    try:
+        shutil.rmtree(opts["cachedir"])
+    except Exception:  # pylint: disable=broad-except
+        # Not my circus, not my monkey
+        pass
```
```suggestion
    shutil.rmtree(opts["cachedir"], ignore_errors=True)
```
?

waynew replied:
:+1:

s0undt3ch commented on tests/pytests/functional/cache/test_cache.py:
```diff
@@ -0,0 +1,526 @@
+    # and force the alter_table to be called. Resetting the client to `None` is
+    # what triggers the implementation to allow the ALTER TABLE to add the
+    # last_update column
+    run_query = cache.modules["mysql.run_query"]
+    run_query(
+        conn=None,
+        query="ALTER TABLE salt_cache.cache DROP COLUMN last_update",
+    )[0].fetchone()
```
Mark as destructive tests?
But isn't mysql running in a container here?

waynew replied:
Well, it shouldn't need to be destructive, because containers.

The problem that this is testing is that *existing* cache users won't have the column on the table. And since when we first connect we're going to create the table *with* the column, I needed a way to test that we'd also alter existing tables.

Which this will do (and is called out in the changelog)

s0undt3ch replied:
But if the test runs in a container, we start, we're not destroying the system the test suite is running on...

s0undt3ch reviewed (changes requested):
As for the consistency questions, we should strive to be consistent across backends...

waynew committed:
incorporate PR feedback

Ch3LL reviewed (commented):
just one question

waynew committed:
Merge branch 'master' into tc/cache_tests

waynew committed:
Skip docker if not found and add the rest of mysql

I really really hate the way this is configured but I can't make pytest
use fixtures as parameters in fixtures. I may have to struggle to come
up with another way.

waynew committed:
Tests all passing, woo!

waynew committed:
Merge remote-tracking branch 'salt/master' into tc/cache_tests

waynew committed:
run black

Ch3LL commented:
lint is failing

waynew committed:
Merge branch 'master' into tc/cache_tests

waynew committed:
Disable pylint here

waynew committed:
Merge remote-tracking branch 'wayne/tc/cache_tests' into tc/cache_tests

waynew committed:
Skip when mysql fails during CI

This has been tested locally and it runs all of them. I'm sure with the
CI environment it's possible that it fails for no particularly good
reason. And it's definitely not worth blocking a PR for (though arguably
it could be if *all* of the images fail, but...)

waynew commented:
Lint should be good now, it was failing on one test for flakiness - just pushed a fix for that. The other one the whole centos 8 stream failed to even launch :shrug: 

We'll see if they all pass now...
